[metadata] Implement LPARRAY to ARRAY marshalling.

rbernon opened the pull request:
I initially intended to try upstreaming this, but it looks like we have non-upstream changes in the area already. Was there any previous attempt?

rbernon committed:
[metadata] Implement LPARRAY to ARRAY marshalling.

rbernon committed:
[metadata] Add more LPARRAY to ARRAY marshalling tests.

rbernon commented:
I think it was leaking refs, I added a call to `Release` after `GetObjectForIUnknown`. I'm not sure if it should release it too when castclass fails... though it's not supposed to happen?

madewokherd commented:
I haven't looked at the code yet, but it's not obvious to me why we would need to Release at all. Objects probably aren't passed in with the expectation that we release them. The garbage collector will take care of releasing the reference created by GetObjectForIUnknown.

madewokherd commented on mono/metadata/marshal-ilgen.c:
```diff
@@ -2700,11 +2700,32 @@ emit_marshal_array_ilgen (EmitMarshalContext *m, int argnum, MonoType *t,
+				mono_mb_emit_byte (mb, CEE_CONV_OVF_I);
+			}
+
+			mono_mb_emit_icall_id (mb, conv_to_icall (MONO_MARSHAL_CONV_LPARRAY_ARRAY, NULL));
```
Why doesn't the part in `if (need_convert || need_free) {` in the case already handle the conversion? Can it be fixed to include the case you're seeing?

rbernon replied:
Well that case seems to be for string and string builder objects:
* need_convert is only set for unicode char arrays,
* need_free is set for stringbuilder class instances.

As I see it, the `MARSHAL_ACTION_CONV_IN` side handles the marshalling of `ARRAY` to `LPARRAY` through `mono_array_to_lparray_impl`, and it seems much more natural to handle the `MARSHAL_ACTION_CONV_OUT` marshalling symmetrically in the function that was only freeing the `LPARRAY`.

Also as it's executed at runtime it's simpler than having to emit IR to do the same.

madewokherd replied:
If I'm reading this right, `need_convert` is true if there's an `OutAttribute` on the parameter (maybe it should be using `is_out` instead).

rbernon replied:
Eh indeed, I got lost in the parentheses sorry.

rbernon replied:
So the reason why that case isn't handled is because the elements COM wrapper class have `native_size` == 0, which means `emit_struct_conv_full` does nothing.

Maybe it could be fixed there, but as the `IN` side does the marshaling through `MONO_MARSHAL_CONV_ARRAY_LPARRAY` (in the `if (m_class_is_blittable (eklass))` case), it seems completely logical to have a similar case on the `OUT` side. There's actually a FIXME comment to indicate that it should be done like that.

madewokherd commented:
The cominterop test failure is likely caused by the PR, the others are from a Wine regression.

rbernon commented:
Yeah looks like it doesn't need to release objects after all.

madewokherd commented:
It'd be nice to have a test for this, though it doesn't have to be in the PR itself. I can add a stand-alone test exe to wine-mono, and it's allowed to depend on a C library or win32 api's.

madewokherd commented:
Sorry, when I said "it'd be nice" what I meant was "I'm going to need a test for this".

rbernon commented:
I added a test, and also simplified a bit the implementation, trying to remove the need for the size parameter and assuming that checking the objects was enough.

Now that I'm considering it again I probably simplified it too much, as there's no guarantee that the end of the array would be cleared or even that we should try to read whatever is there. I'll do another update.

rbernon committed:
[metadata] Add a helper to emit Out LPARRAY size load.

rbernon commented:
Updated with more tests. There's still some small differences with native with how Out only LPArray marshaling is done, where native clears the array entirely, while mono doesn't. I'm not sure how to pass the marshaling attribute to the conversion function, and I don't think it matters too much anyway.

The Out marshaling doesn't seem to really use the output array size, except when the array is passed by ref, in which case it uses it to create the new array. I fixed the handling of combined SizeConst and SizeParamIndex in that case.

Otherwise the entire array is converted back, and if native has put some invalid pointer values there, including after the returned size in the non-byref case, it just crashes, so I assume there's no need to pass the size to the conversion function. I added special handling for NULL values as GetObjectForIUnknown would throw otherwise.

rbernon commented:
I added a fix for the pinvoke2 test failure, the issue was that Out marshaling with basic types was not copying data from native array to managed array. The managed array may have been reallocated, when it is passed byref, and data needs to be copied in that case.

rbernon commented:
From the windows and linux test logs it seems like the remaining failures are from some network test that now fails on both?

madewokherd commented:
I haven't looked yet, but there have been network tests acting up lately, so that seems likely. I'll let you know if there's anything in there that looks relevant.
